File: LeetCode/348.Design_Tic-Tac-Toe/Mario/348.Design_Tic-Tac-Toe.py
#!/usr/bin/env python3


class TicTacToe:

    def __init__(self, n: int):
        self.n = n
        self.rows = [0] * n
        self.cols = [0] * n
        self.diag = 0
        self.anti_diag = 0

    def move(self, row: int, col: int, player: int) -> int:
        # player 1 add 1, player 2 add -1
        val = 1 if player == 1 else -1

        self.rows[row] += val
        self.cols[col] += val

        # diagonal
        if row == col:
            self.diag += val

        # anti_diagonal
        if row + col == self.n - 1:
            self.anti_diag += val

        if (
            abs(self.rows[row]) == self.n
            or abs(self.cols[col]) == self.n
            or abs(self.diag) == self.n
            or abs(self.anti_diag) == self.n
        ):
            return player

        return 0

File: LeetCode/348.Design_Tic-Tac-Toe/348.Design_Tic-Tac-Toe.swift
class TicTacToe {
    // rows, if play 1, rows[row] += 1, play2, -1
    var rows: [Int], cols: [Int], diag: [Int]
    let n: Int

    init(_ n: Int) {
        self.n = n
        rows = Array(repeating: 0, count: n)
        cols = Array(repeating: 0, count: n)
        diag = [0, 0] // diag, antidiag
    }

    func move(_ row: Int, _ col: Int, _ player: Int) -> Int {
        var add = player == 1 ? 1 : -1
        rows[row] += add
        cols[col] += add
        if row == col { diag[0] += add }
        if row + col == n - 1 { diag[1] += add }

        if abs(rows[row]) == n || abs(cols[col]) == n || abs(diag[0]) == n || abs(diag[1]) == n { return player }
        return 0
    }
}
